pkg/monitor/controller_test.go

```diff
@@ -0,0 +1,719 @@
+			remainingAgentCRDs: []string{"testAgentCRD1"},
+		},
+	}
+	for _, tt := range tc {
+		t.Run(tt.name, func(t *testing.T) {
+			clientset := fakeclientset.NewSimpleClientset()
+			if tt.existingExternalNode != nil {
+				clientset = fakeclientset.NewSimpleClientset(tt.existingExternalNode)
+			}
+			tt.prepareReactor(clientset)
+			controller := newControllerMonitor(clientset)
+			stopCh := make(chan struct{})
+			initController(controller, stopCh)
+			defer close(stopCh)
+			controller.controllerMonitor.externalNodeEnabled = true
+			for _, crd := range tt.agentCRDs {
+				agentCRD := new(v1beta1.AntreaAgentInfo)
+				agentCRD.Name = crd
+				agentCRD.ResourceVersion = "0"
+				_, err := controller.controllerMonitor.client.CrdV1beta1().AntreaAgentInfos().Create(ctx, agentCRD, metav1.CreateOptions{})
+				assert.NoError(t, err)
+			}
+			controller.controllerMonitor.deleteStaleAgentCRDs()
+			crds, err := controller.controllerMonitor.client.CrdV1beta1().AntreaAgentInfos().List(ctx, metav1.ListOptions{ResourceVersion: "0"})
+			if tt.expectedError != "" {
+				assert.Equal(t, tt.expectedError, err.Error())
+			} else {
+				require.NoError(t, err)
+				remainingAgentCRDs := []string{}
+				for _, crd := range crds.Items {
+					remainingAgentCRDs = append(remainingAgentCRDs, crd.Name)
+				}
+				assert.Equal(t, tt.remainingAgentCRDs, remainingAgentCRDs)
+			}
+		})
+	}
+}
+
+func TestSyncControllerCRD(t *testing.T) {
+	existingCRD := &v1beta1.AntreaControllerInfo{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: crdName,
+		},
+		NetworkPolicyControllerInfo: v1beta1.NetworkPolicyControllerInfo{
+			NetworkPolicyNum: 1,
+		},
+		APIPort: 0,
+	}
+	partiallyUpdatedCRD := &v1beta1.AntreaControllerInfo{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: crdName,
+		},
+		NetworkPolicyControllerInfo: v1beta1.NetworkPolicyControllerInfo{
+			NetworkPolicyNum: 0,
+		},
+		APIPort: 0,
+	}
+	entirelyUpdatedCRD := &v1beta1.AntreaControllerInfo{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: crdName,
+		},
+		NetworkPolicyControllerInfo: v1beta1.NetworkPolicyControllerInfo{
+			NetworkPolicyNum: 0,
+		},
+		APIPort: 10349,
+	}
+	newCRD := &v1beta1.AntreaControllerInfo{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: crdName,
+		},
+		Version: "UNKNOWN",
+		NodeRef: v1.ObjectReference{
+			Kind: "Node",
+		},
+		ServiceRef: v1.ObjectReference{
+			Kind: "Service",
+			Name: "antrea",
+		},
+		NetworkPolicyControllerInfo: v1beta1.NetworkPolicyControllerInfo{
+			NetworkPolicyNum:  0,
+			AddressGroupNum:   0,
+			AppliedToGroupNum: 0,
+		},
+		APIPort: 10349,
+	}
+	t.Run("partial update-success", func(t *testing.T) {
+		clientset := fakeclientset.NewSimpleClientset(existingCRD)
+		controller := newControllerMonitor(clientset)
+		stopCh := make(chan struct{})
+		initController(controller, stopCh)
+		defer close(stopCh)
+		controller.controllerMonitor.controllerCRD = existingCRD
+		controller.controllerMonitor.querier.GetControllerInfo(partiallyUpdatedCRD, true)
```

Comment:
why are you calling `GetControllerInfo` yourself in the test, I don't understand? Isn't `syncControllerCRD` supposed to call it?